### ISSUE TITLE
gradient check failed

### ISSUE THREAD

**Issue by zcyang** [cvxgrp/cvxpylayers#67]
I am testing the gradient of cvxpylayer using finite difference method. Below is an example script:
```
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.autograd import Variable
from torch.nn.parameter import Parameter

import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer

N = 100
M = 100
NORM_BOUND = 1

class Opt(nn.Module):
  def __init__(self):
    super(Opt, self).__init__()
    self.L = torch.randn(N, M)

  def forward(self, x, weight):
    xw = x.mm(weight).squeeze(-1)

    _p = cp.Parameter((N))
    _v = cp.Variable((N))
    obj = cp.Minimize(0.5*cp.sum_squares(self.L.T@_v)
                      + 0.5*cp.sum_squares(_v)
                      - _v@_p)
    cons = [_v <= 1, -_v <= 1, cp.norm(_v, 1) <= NORM_BOUND]
    prob = cp.Problem(obj, cons)
    self.cvxopt = CvxpyLayer(prob, [_p], [_v])

    v, = self.cvxopt(xw)
    return v

def main():
  opt = Opt()

  x = torch.randn(N, 2)
  weight = torch.tensor([[0.5], [0.5]], requires_grad=True)
  weight1 = torch.tensor([[0.5+1e-5], [0.5]], requires_grad=False)
  weight2 = torch.tensor([[0.5], [0.5+1e-5]], requires_grad=False)
  y = torch.randn(N)

  v = opt(x, weight)
  v1 = opt(x, weight1)
  v2 = opt(x, weight2)
  loss = torch.sum(y*v)
  loss1 = torch.sum(y*v1)
  loss2 = torch.sum(y*v2)

  loss.backward()
  print(weight.grad)
  print((loss1-loss)/1e-5)
  print((loss2-loss)/1e-5)
```

The gradient returned through backward is
```
tensor([[ 0.1767],
   [-0.1480]])
```
while the gradient with finite different is:
```
tensor(-7.9798, grad_fn=<DivBackward0>)
tensor(-5.3742, grad_fn=<DivBackward0>)
```

The gap is very large...

The gap is still large if I remove the norm constraint
```
    cons = [_v <= 0.1, -_v <= 0.1]
```

If all the constraints are removed
```
cons = []
```
the gap seems to be much smaller, but still large sometimes.

**Comment by akshayka** [comment]
For your particular problem, you need higher accuracy solves --- it appears that the optimal value for v has very small entries, which makes the numerical gradient unstable. If you replace 

```python3
v, = self.cvxopt(xw)
```

with

```python3
v, = self.cvxopt(xw, solver_args={'eps': 1e-8,  'max_iters': 2000})
```

and set the random seed to 0 at the beginning of main (`torch.manual_seed(0)`). (The 'max_iters' change isn't necessary here, but you may find it useful in the future.)

You should see something like the following output, which appears reasonable to me:

```python3
tensor([[ 0.0464],
        [-0.1543]])
tensor(0.0343, grad_fn=<DivBackward0>)
tensor(-0.1118, grad_fn=<DivBackward0>)
```

By the way, you likely should create the layer once, during initialization. Recreating the layer in the forward pass every time will likely be much slower than just creating it once and re-using it thereafter.

Full code:
```python3
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.autograd import Variable
from torch.nn.parameter import Parameter

import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer

N = 100
M = 100
NORM_BOUND = 1

class Opt(nn.Module):
  def __init__(self):
    super(Opt, self).__init__()
    self.L = torch.randn(N, M)
    _p = cp.Parameter((N))
    _v = cp.Variable((N))
    obj = cp.Minimize(0.5*cp.sum_squares(self.L.T@_v)
                      + 0.5*cp.sum_squares(_v)
                      - _v@_p)
    cons = [_v <= 1, -_v <= 1, cp.norm(_v, 1) <= NORM_BOUND]
    prob = cp.Problem(obj, cons)
    self.cvxopt = CvxpyLayer(prob, [_p], [_v])

  def forward(self, x, weight):
    xw = x.mm(weight).squeeze(-1)
    v, = self.cvxopt(xw, solver_args={'eps': 1e-8,  'max_iters':2000})
    return v

def main():
  torch.manual_seed(0)
  opt = Opt()

  x = torch.randn(N, 2)
  weight = torch.tensor([[0.5], [0.5]], requires_grad=True)
  weight1 = torch.tensor([[0.5+1e-5], [0.5]], requires_grad=False)
  weight2 = torch.tensor([[0.5], [0.5+1e-5]], requires_grad=False)
  y = torch.randn(N)

  v = opt(x, weight)
  v1 = opt(x, weight1)
  v2 = opt(x, weight2)
  loss = torch.sum(y*v)
  loss1 = torch.sum(y*v1)
  loss2 = torch.sum(y*v2)

  loss.backward()
  print(weight.grad)
  print((loss1-loss)/1e-5)
  print((loss2-loss)/1e-5)
```

**Comment by zcyang** [comment]
Thanks for the reply! 

I met a problem when I was trying to overfit a single training data point. But training loss has been steadily increasing:

```
epoch 0 itr 0 loss -35.929
epoch 1 itr 0 loss -35.907
epoch 2 itr 0 loss -35.884
epoch 3 itr 0 loss -35.862
epoch 4 itr 0 loss -35.839
epoch 5 itr 0 loss -35.817
epoch 6 itr 0 loss -35.794
epoch 7 itr 0 loss -35.771
epoch 8 itr 0 loss -35.749
epoch 9 itr 0 loss -35.726
epoch 10 itr 0 loss -35.704
epoch 11 itr 0 loss -35.682
epoch 12 itr 0 loss -35.660
epoch 13 itr 0 loss -35.638
epoch 14 itr 0 loss -35.616
epoch 15 itr 0 loss -35.594
epoch 16 itr 0 loss -35.572
epoch 17 itr 0 loss -35.550
epoch 18 itr 0 loss -35.529
epoch 19 itr 0 loss -35.507
epoch 20 itr 0 loss -35.486
epoch 21 itr 0 loss -35.465
epoch 22 itr 0 loss -35.443
epoch 23 itr 0 loss -35.421
epoch 24 itr 0 loss -35.400
epoch 25 itr 0 loss -35.378
epoch 26 itr 0 loss -35.357
epoch 27 itr 0 loss -35.336
epoch 28 itr 0 loss -35.314
epoch 29 itr 0 loss -35.293
epoch 30 itr 0 loss -35.272
epoch 31 itr 0 loss -35.251
epoch 32 itr 0 loss -35.230
epoch 33 itr 0 loss -35.209
```

I was suspecting there could be some problems with the gradient.
I tried to set `eps=1e-8` and the problem remained.  

Do you have any idea what could be the cause?

Thanks!
